File: backend/app/api/obligations.py
from datetime import date, datetime

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from backend.app.database import get_db
from backend.app.models.obligation import Obligation
from backend.app.models.contract import Contract
from backend.app.models.user import User
from backend.app.schemas.obligation import (
    ObligationCreate,
    ObligationUpdate,
    ObligationStatusUpdate,
    ObligationOut,
)
from backend.app.core.auth import get_current_user


router = APIRouter()


# ============================================================
# CREATE OBLIGATION
# ============================================================

@router.post(
    "/",
    response_model=ObligationOut,
    status_code=201
)
def create_obligation(
    obligation_data: ObligationCreate,
    db: Session = Depends(get_db),
    current_user: dict = Depends(get_current_user)
):

    # Check contract
    contract = db.query(Contract).filter(
        Contract.id == obligation_data.contract_id
    ).first()

    if not contract:
        raise HTTPException(
            status_code=404,
            detail="Contract not found"
        )

    # Check assigned user
    user = db.query(User).filter(
        User.id == obligation_data.assigned_to
    ).first()

    if not user:
        raise HTTPException(
            status_code=404,
            detail="Assigned user not found"
        )

    # Create obligation
    new_obligation = Obligation(
        contract_id=obligation_data.contract_id,
        title=obligation_data.title,
        description=obligation_data.description,
        obligation_type=obligation_data.obligation_type,
        due_date=obligation_data.due_date,
        assigned_to=obligation_data.assigned_to,
        status="Pending"
    )

    db.add(new_obligation)
    db.commit()
    db.refresh(new_obligation)

    return new_obligation


# ============================================================
# GET ALL OBLIGATIONS
# ============================================================

@router.get(
    "/",
    response_model=list[ObligationOut]
)
def get_obligations(
    db: Session = Depends(get_db),
    current_user: dict = Depends(get_current_user)
):

    obligations = db.query(Obligation).all()

    return obligations


# ============================================================
# GET OBLIGATIONS FOR CONTRACT
# ============================================================

@router.get(
    "/contracts/{contract_id}/obligations",
    response_model=list[ObligationOut]
)
def get_contract_obligations(
    contract_id: int,
    db: Session = Depends(get_db),
    current_user: dict = Depends(get_current_user)
):

    # Check contract
    contract = db.query(Contract).filter(
        Contract.id == contract_id
    ).first()

    if not contract:
        raise HTTPException(
            status_code=404,
            detail="Contract not found"
        )

    obligations = db.query(Obligation).filter(
        Obligation.contract_id == contract_id
    ).all()

    return obligations


# ============================================================
# UPDATE OBLIGATION
# ============================================================

@router.put(
    "/{obligation_id}",
    response_model=ObligationOut
)
def update_obligation(
    obligation_id: int,
    obligation_data: ObligationUpdate,
    db: Session = Depends(get_db),
    current_user: dict = Depends(get_current_user)
):

    obligation = db.query(Obligation).filter(
        Obligation.id == obligation_id
    ).first()

    if not obligation:
        raise HTTPException(
            status_code=404,
            detail="Obligation not found"
        )

    # Check assigned user if changing assignment
    if obligation_data.assigned_to is not None:

        user = db.query(User).filter(
            User.id == obligation_data.assigned_to
        ).first()

        if not user:
            raise HTTPException(
                status_code=404,
                detail="Assigned user not found"
            )

        obligation.assigned_to = obligation_data.assigned_to

    if obligation_data.title is not None:
        obligation.title = obligation_data.title

    if obligation_data.description is not None:
        obligation.description = obligation_data.description

    if obligation_data.obligation_type is not None:
        obligation.obligation_type = obligation_data.obligation_type

    if obligation_data.due_date is not None:
        obligation.due_date = obligation_data.due_date

    db.commit()
    db.refresh(obligation)

    return obligation


# ============================================================
# UPDATE OBLIGATION STATUS
# ============================================================

@router.patch(
    "/{obligation_id}/status",
    response_model=ObligationOut
)
def update_obligation_status(
    obligation_id: int,
    status_data: ObligationStatusUpdate,
    db: Session = Depends(get_db),
    current_user: dict = Depends(get_current_user)
):

    obligation = db.query(Obligation).filter(
        Obligation.id == obligation_id
    ).first()

    if not obligation:
        raise HTTPException(
            status_code=404,
            detail="Obligation not found"
        )

    allowed_statuses = [
        "Pending",
        "In Progress",
        "Completed",
        "Delayed",
        "Overdue"
    ]

    if status_data.status not in allowed_statuses:
        raise HTTPException(
            status_code=400,
            detail="Invalid obligation status"
        )

    current_status = obligation.status
    new_status = status_data.status

    valid_transitions = {
        "Pending": ["In Progress", "Overdue"],
        "In Progress": ["Completed", "Delayed", "Overdue"],
        "Delayed": ["In Progress", "Completed", "Overdue"],
        "Overdue": ["In Progress", "Completed"],
        "Completed": []
    }

    if new_status not in valid_transitions.get(current_status, []):
        raise HTTPException(
            status_code=400,
            detail=f"Invalid status transition from {current_status} to {new_status}"
        )

    obligation.status = new_status

    if new_status == "Completed":
        obligation.completion_date = date.today()

    db.commit()
    db.refresh(obligation)

    return obligation


# ============================================================
# COMPLETE OBLIGATION
# ============================================================

@router.post(
    "/{obligation_id}/complete",
    response_model=ObligationOut
)
def complete_obligation(
    obligation_id: int,
    db: Session = Depends(get_db),
    current_user: dict = Depends(get_current_user)
):

    obligation = db.query(Obligation).filter(
        Obligation.id == obligation_id
    ).first()

    if not obligation:
        raise HTTPException(
            status_code=404,
            detail="Obligation not found"
        )

    if obligation.status == "Completed":
        raise HTTPException(
            status_code=400,
            detail="Obligation is already completed"
        )

    obligation.status = "Completed"
    obligation.completion_date = date.today()

    db.commit()
    db.refresh(obligation)

    return obligation


# ============================================================
# OVERDUE OBLIGATIONS
# ============================================================

@router.get(
    "/overdue/list",
    response_model=list[ObligationOut]
)
def get_overdue_obligations(
    db: Session = Depends(get_db),
    current_user: dict = Depends(get_current_user)
):

    today = date.today()

    obligations = db.query(Obligation).filter(
        Obligation.due_date < today,
        Obligation.status != "Completed"
    ).all()

    return obligations
#==============================================================
# GET BY ID 
#==============================================================

@router.get(
    "/{obligation_id}",
    response_model=ObligationOut
)
def get_obligation(
    obligation_id: int,
    db: Session = Depends(get_db),
    current_user: dict = Depends(get_current_user)
):

    obligation = db.query(Obligation).filter(
        Obligation.id == obligation_id
    ).first()

    if not obligation:
        raise HTTPException(
            status_code=404,
            detail="Obligation not found"
        )

    return obligation

File: src/app/obligations/obligations.ts
import { CommonModule } from '@angular/common';
import {
  ChangeDetectorRef,
  Component,
  OnInit
} from '@angular/core';
import { FormsModule } from '@angular/forms';

import { MatButtonModule } from '@angular/material/button';
import { MatCardModule } from '@angular/material/card';
import { MatDialogModule } from '@angular/material/dialog';
import { MatFormFieldModule } from '@angular/material/form-field';
import { MatInputModule } from '@angular/material/input';
import { MatSelectModule } from '@angular/material/select';
import { MatIconModule } from '@angular/material/icon';

import {
  Obligation,
  CreateObligationRequest,
  UpdateObligationRequest
} from '../models/obligation.model';

import { ObligationService } from '../services/obligation.service';

@Component({
  selector: 'app-obligations',
  standalone: true,
  imports: [
    CommonModule,
    FormsModule,
    MatButtonModule,
    MatCardModule,
    MatDialogModule,
    MatFormFieldModule,
    MatInputModule,
    MatSelectModule,
    MatIconModule
  ],
  templateUrl: './obligations.html',
  styleUrl: './obligations.css'
})
export class Obligations implements OnInit {
  obligations: Obligation[] = [];

  filteredObligations: Obligation[] = [];

  searchTerm = '';
  selectedStatus = 'All';

  isLoading = false;
  isSaving = false;

  errorMessage = '';
  successMessage = '';

  showForm = false;
  editingObligationId: number | null = null;

  formData: CreateObligationRequest = {
    contract_id: 0,
    title: '',
    description: '',
    obligation_type: '',
    due_date: '',
    assigned_to: null
  };

  statusOptions = [
    'Pending',
    'Completed',
    'Overdue',
    'In Progress'
  ];

  constructor(
    private obligationService: ObligationService,
    private cdr: ChangeDetectorRef
  ) {}

  ngOnInit(): void {
    this.loadObligations();
  }

  loadObligations(): void {
    this.isLoading = true;
    this.errorMessage = '';

    this.cdr.detectChanges();

    this.obligationService.getObligations().subscribe({
      next: (data) => {
        this.obligations = Array.isArray(data) ? data : [];

        this.applyFilters();

        this.isLoading = false;

        // Force the page to refresh immediately after API data arrives.
        this.cdr.detectChanges();

        console.log(
          'Obligations loaded successfully:',
          this.obligations
        );
        console.log(
          'Filtered obligations:',
          this.filteredObligations
        );
      },

      error: (error) => {
        console.error(
          'Failed to load obligations:',
          error
        );

        this.obligations = [];
        this.filteredObligations = [];

        this.errorMessage =
          this.getErrorMessage(
            error,
            'Unable to load obligations.'
          );

        this.isLoading = false;

        this.cdr.detectChanges();
      }
    });
  }

  applyFilters(): void {
    const search = this.searchTerm
      .trim()
      .toLowerCase();

    this.filteredObligations =
      this.obligations.filter(
        (obligation) => {
          const title =
            String(obligation.title ?? '')
              .toLowerCase();

          const obligationType =
            String(
              obligation.obligation_type ?? ''
            ).toLowerCase();

          const contractId =
            String(obligation.contract_id ?? '');

          const matchesSearch =
            !search ||
            title.includes(search) ||
            obligationType.includes(search) ||
            contractId.includes(search);

          const matchesStatus =
            this.selectedStatus === 'All' ||
            obligation.status === this.selectedStatus;

          return (
            matchesSearch &&
            matchesStatus
          );
        }
      );
  }

  onSearchChange(value: string): void {
    this.searchTerm = value ?? '';

    this.applyFilters();

    this.cdr.detectChanges();
  }

  onStatusFilterChange(value: string): void {
    this.selectedStatus = value ?? 'All';

    this.applyFilters();

    this.cdr.detectChanges();
  }

  openCreateForm(): void {
    this.resetForm();

    this.showForm = true;

    this.errorMessage = '';
    this.successMessage = '';

    this.cdr.detectChanges();
  }

  openEditForm(
    obligation: Obligation
  ): void {
    this.editingObligationId =
      obligation.id;

    this.formData = {
      contract_id: obligation.contract_id,
      title: obligation.title,
      description:
        obligation.description ?? '',
      obligation_type:
        obligation.obligation_type,
      due_date: obligation.due_date,

      // assigned_to is number | null
      assigned_to:
        obligation.assigned_to ?? null
    };

    this.showForm = true;
    this.successMessage = '';
    this.errorMessage = '';

    this.cdr.detectChanges();
  }

  closeForm(): void {
    if (this.isSaving) {
      return;
    }

    this.showForm = false;
    this.editingObligationId = null;

    this.resetForm();

    this.cdr.detectChanges();
  }

  saveObligation(): void {
    this.errorMessage = '';
    this.successMessage = '';

    if (!this.isValidForm()) {
      this.errorMessage =
        'Please fill in all required fields.';

      this.cdr.detectChanges();

      return;
    }

    this.isSaving = true;

    this.cdr.detectChanges();

    /*
     * CREATE
     */
    if (
      this.editingObligationId === null
    ) {
      const payload: CreateObligationRequest = {
        contract_id:
          Number(this.formData.contract_id),

        title:
          this.formData.title.trim(),

        description:
          this.formData.description?.trim() ?? '',

        obligation_type:
          this.formData.obligation_type.trim(),

        due_date:
          this.formData.due_date,

        /*
         * FIX:
         * assigned_to is number | null.
         * Do not compare it with ''.
         */
        assigned_to:
          this.formData.assigned_to ?? null
      };

      this.obligationService
        .createObligation(payload)
        .subscribe({
          next: (created) => {
            this.obligations = [
              created,
              ...this.obligations
            ];

            this.applyFilters();

            this.successMessage =
              'Obligation created successfully.';

            this.isSaving = false;
            this.showForm = false;

            this.resetForm();

            this.cdr.detectChanges();
          },

          error: (error) => {
            console.error(
              'Failed to create obligation:',
              error
            );

            this.errorMessage =
              this.getErrorMessage(
                error,
                'Unable to create obligation.'
              );

            this.isSaving = false;

            this.cdr.detectChanges();
          }
        });

      return;
    }

    /*
     * UPDATE
     */
    const updatePayload:
      UpdateObligationRequest = {
      title:
        this.formData.title.trim(),

      description:
        this.formData.description?.trim() ?? '',

      obligation_type:
        this.formData.obligation_type.trim(),

      due_date:
        this.formData.due_date,

      /*
       * FIX:
       * assigned_to is number | null.
       * Do not compare it with ''.
       */
      assigned_to:
        this.formData.assigned_to ?? null
    };

    this.obligationService
      .updateObligation(
        this.editingObligationId,
        updatePayload
      )
      .subscribe({
        next: (updated) => {
          this.obligations =
            this.obligations.map(
              (item) =>
                item.id === updated.id
                  ? updated
                  : item
            );

          this.applyFilters();

          this.successMessage =
            'Obligation updated successfully.';

          this.isSaving = false;
          this.showForm = false;
          this.editingObligationId = null;

          this.resetForm();

          this.cdr.detectChanges();
        },

        error: (error) => {
          console.error(
            'Failed to update obligation:',
            error
          );

          this.errorMessage =
            this.getErrorMessage(
              error,
              'Unable to update obligation.'
            );

          this.isSaving = false;

          this.cdr.detectChanges();
        }
      });
  }

  viewObligation(
    obligation: Obligation
  ): void {
    this.errorMessage = '';

    this.obligationService
      .getObligation(obligation.id)
      .subscribe({
        next: (details) => {
          window.alert(
            [
              `ID: ${details.id}`,
              `Title: ${details.title}`,
              `Contract ID: ${details.contract_id}`,
              `Type: ${details.obligation_type}`,
              `Due Date: ${details.due_date}`,
              `Status: ${details.status}`,
              `Assigned To: ${
                details.assigned_to ??
                'Not assigned'
              }`,
              `Description: ${
                details.description ??
                'No description'
              }`
            ].join('\n')
          );

          this.cdr.detectChanges();
        },

        error: (error) => {
          console.error(
            'Failed to load obligation details:',
            error
          );

          this.errorMessage =
            this.getErrorMessage(
              error,
              'Unable to load obligation details.'
            );

          this.cdr.detectChanges();
        }
      });
  }

  changeStatus(
    obligation: Obligation,
    status: string
  ): void {
    if (
      !status ||
      status === obligation.status
    ) {
      return;
    }

    this.errorMessage = '';
    this.successMessage = '';

    this.obligationService
      .updateStatus(
        obligation.id,
        { status }
      )
      .subscribe({
        next: (updated) => {
          this.obligations =
            this.obligations.map(
              (item) =>
                item.id === updated.id
                  ? updated
                  : item
            );

          this.applyFilters();

          this.successMessage =
            'Obligation status updated successfully.';

          this.cdr.detectChanges();
        },

        error: (error) => {
          console.error(
            'Failed to update obligation status:',
            error
          );

          this.errorMessage =
            this.getErrorMessage(
              error,
              'Unable to update obligation status.'
            );

          this.cdr.detectChanges();
        }
      });
  }

  completeObligation(
    obligation: Obligation
  ): void {
    if (
      obligation.status === 'Completed'
    ) {
      return;
    }

    this.errorMessage = '';
    this.successMessage = '';

    this.obligationService
      .completeObligation(obligation.id)
      .subscribe({
        next: (updated) => {
          this.obligations =
            this.obligations.map(
              (item) =>
                item.id === updated.id
                  ? updated
                  : item
            );

          this.applyFilters();

          this.successMessage =
            'Obligation marked as completed.';

          this.cdr.detectChanges();
        },

        error: (error) => {
          console.error(
            'Failed to complete obligation:',
            error
          );

          this.errorMessage =
            this.getErrorMessage(
              error,
              'Unable to complete obligation.'
            );

          this.cdr.detectChanges();
        }
      });
  }

  getStatusClass(
    status: string
  ): string {
    return String(status ?? '')
      .toLowerCase()
      .replace(/\s+/g, '-');
  }

  isOverdue(
    obligation: Obligation
  ): boolean {
    if (
      obligation.status === 'Completed'
    ) {
      return false;
    }

    if (!obligation.due_date) {
      return false;
    }

    const dueDate = new Date(
      `${obligation.due_date}T00:00:00`
    );

    const today = new Date();

    today.setHours(
      0,
      0,
      0,
      0
    );

    return dueDate < today;
  }

  private isValidForm(): boolean {
    return (
      Number(this.formData.contract_id) > 0 &&
      this.formData.title.trim().length > 0 &&
      this.formData.obligation_type.trim().length > 0 &&
      this.formData.due_date.length > 0
    );
  }

  private resetForm(): void {
    this.formData = {
      contract_id: 0,
      title: '',
      description: '',
      obligation_type: '',
      due_date: '',
      assigned_to: null
    };

    this.editingObligationId = null;
    this.isSaving = false;
  }

  private getErrorMessage(
    error: any,
    fallback: string
  ): string {
    if (error?.status === 401) {
      return (
        'Your session is not authorized. ' +
        'Please log in again.'
      );
    }

    if (error?.status === 403) {
      return (
        'You are not authorized to perform this action.'
      );
    }

    if (error?.status === 404) {
      return (
        'The requested obligation was not found.'
      );
    }

    if (error?.status === 422) {
      return (
        'Some entered data is invalid. ' +
        'Please check the form.'
      );
    }

    if (
      error?.error?.detail &&
      typeof error.error.detail === 'string'
    ) {
      return error.error.detail;
    }

    return fallback;
  }
}